src/box/lua/config/applier/credentials.lua

```diff
@@ -277,85 +282,291 @@ local function privileges_add_defaults(name, role_or_user, intermediate)
     end
 
-    config_privileges = privileges_add_defaults(name, role_or_user,
-                                                config_privileges)
+    -- The privileges synchronization between A and B is performed in 3 steps:
+    -- 1. Grant all privileges that are present in B,
+    --    but not present in A (`grant(B - A)`).
+    -- 2. Add default privileges to B (`B = B + defaults`).
+    -- 3. Revoke all privileges that are not present in B,
+    --    but present in A (`revoke(A - B)).
+    --
+    -- Default privileges are not granted on step 1, so they stay revoked if
+    -- revoked manually (e.g. with `box.schema.{user,role}.revoke()`).
+    -- However, defaults should never be revoked, so target state B is enriched
+    -- with them before step 3.
+    local function sync(role_or_user, name, config_privileges)
+        assert(role_or_user == 'user' or role_or_user == 'role')
+        if not obj_to_sync then
+            log.verbose('syncing privileges for %s %q', role_or_user, name)
+        end
+
+        local box_privileges = box.schema[role_or_user].info(name)
 
-    local revokes = privileges_subtract(box_privileges, config_privileges)
+        config_privileges = privileges_from_config(config_privileges)
+        box_privileges = privileges_from_box(box_privileges)
 
-    for _, to_revoke in ipairs(revokes) do
-        revoke_f(name, to_revoke.privs, to_revoke.obj_type, to_revoke.obj_name)
+        local grants = privileges_subtract(config_privileges, box_privileges)
+
+        for _, to_grant in ipairs(grants) do
+            -- Note that grants are filtered per object, if required.
+            if obj_to_sync == nil or (obj_to_sync.type == to_grant.obj_type and
+                                     obj_to_sync.name == to_grant.obj_name) then
+                privileges_action_f('grant', role_or_user, name, to_grant.privs,
+                                    to_grant.obj_type, to_grant.obj_name)
+            end
+        end
+
+        config_privileges = privileges_add_defaults(name, role_or_user,
+                                                    config_privileges)
+
+        local revokes = privileges_subtract(box_privileges, config_privileges)
+
+        for _, to_revoke in ipairs(revokes) do
+            -- Note that revokes are filtered per object, if required.
+            if obj_to_sync == nil or (obj_to_sync.type == to_revoke.obj_type and
+                                    obj_to_sync.name == to_revoke.obj_name) then
+                privileges_action_f('revoke', role_or_user, name,
+                                    to_revoke.privs, to_revoke.obj_type,
+                                    to_revoke.obj_name)
+            end
+        end
     end
 
+    -- The logic inside the sync() function, i.e. privileges dump from
+    -- box, diff calculation and grants/revokes should happen inside one
+    -- transaction. It is the case when called from set_credentials(). When
```

Comment:
There is no `set_credentials()` function now. It seems, the comment needs an update.